yellowbox/extras/websocket.py

```diff
@@ -0,0 +1,347 @@
+            uri: Optional[str] = None, *,
+            regex: Optional[Union[Pattern[str], str]] = None,
+            _overwrite: bool = False) -> None:
+        """Add a route.
+
+        Raises an exception if the route already exists.
+
+        Args:
+            side_effect: Side effect can be either a string or a
+            generator function.
+
+        Raises:
+            RuntimeError: Trying to add a route while it already exists.
+        """
+        if uri and regex:
+            raise ValueError("Only one of URI or regex can be specified.")
+
+        if not (uri or regex):
+            raise ValueError("URI or regex must be specified.")
+
+        if regex:
+            regex = re.compile(regex)
+        elif not isinstance(uri, str):
+            # Prevent variable order confusion.
+            raise TypeError("URI must be a string.")
+
+        gen = _to_generator(side_effect)
+
+        # Check and place
+        with self._lock:
+            if not _overwrite and (uri in self._routes or
+                                   regex in self._re_routes):
+                raise RuntimeError(f"Route {uri or regex} already exists!")
+
+            if uri:
+                self._routes[uri] = gen
+            else:
+                self._re_routes[regex] = gen  # type: ignore
+
+    @contextmanager
+    def patch(self, side_effect: SIDE_EFFECT_TYPE,
+              uri: Optional[str] = None, *,
+              regex: Optional[Union[Pattern[str], str]] = None) -> Iterator[None]:
+        """Temporarily patch a route.
+
+        Raises an exception if the route already exists.
+
+        Args:
+            Same as add().
+        """
+        self.add(side_effect, uri, regex=regex)
+        try:
+            yield
+        finally:
+            self.remove(uri, regex=regex)
+
+    def set(self, side_effect: SIDE_EFFECT_TYPE,
+            uri: Optional[str] = None, *,
+            regex: Optional[Union[Pattern[str], str]] = None) -> None:
+        """Set a route.
+
+        Like `add()` but overwrites existing routes.
+        """
+        self.add(side_effect=side_effect, uri=uri,
+                 regex=regex, _overwrite=True)
+
+    def remove(self, uri: Optional[str] = None, *,
+               regex: Optional[Union[Pattern[str], str]] = None) -> None:
+        """Remove a route."""
+        if uri and regex:
+            raise ValueError("Only one of URI or regex can be specified.")
+
+        if not (uri or regex):
+            raise ValueError("URI or regex must be specified.")
+
+        if uri:
+            del self._routes[uri]
+        else:
+            assert regex  # For Mypy
+            regex = re.compile(regex)
+            with self._lock:
+                del self._re_routes[regex]
+
+    def clear(self) -> None:
+        """Remove all routes."""
+        with self._lock:
+            self._routes.clear()
+            self._re_routes.clear()
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO)
```

Comment:
why is this necessary?

Reply:
Made it for myself mostly as an easy server for testing and thought I'd leave it as is. Proved to be quite useful.

Reply:
It should be deleted, i think

Reply:
why? 'tis echo server in websocket.
can be good for testing and vewy comfowtable. you no think?

Reply:
I don't think we want to encourage people running our modules as mains